library/src/scala/internal/quoted/Matcher.scala

```diff
@@ -277,6 +301,31 @@ private[quoted] object Matcher {
       }
     end treeOps
 
+    private object ClosedTerm {
+      def unapply(term: Term)(given Context, Env): Option[term.type] =
+        if freeVars(term).isEmpty then Some(term) else None
+
+      def freeVars(tree: Tree)(given qctx: Context, env: Env): Set[Symbol] =
+        val accumulator = new TreeAccumulator[Set[Symbol]] {
+          def foldTree(x: Set[Symbol], tree: Tree)(given ctx: Context): Set[Symbol] =
+            tree match
+              case tree: Ident if env.contains(tree.symbol) => foldOverTree(x + tree.symbol, tree)
+              case _ => foldOverTree(x, tree)
+        }
+        accumulator.foldTree(Set.empty, tree)
+    }
+
+    private object IdentArgs {
+      def unapply(args: List[Term])(given Context): Option[List[Ident]] =
+        args.foldRight(Option(List.empty[Ident])) {
+          case (id: Ident, Some(acc)) => Some(id :: acc)
+          case (Block(List(DefDef("$anonfun", Nil, List(params), Inferred(), Some(Apply(id: Ident, args)))), Closure(Ident("$anonfun"), None)), Some(acc))
+              if params.zip(args).forall(_.symbol == _.symbol) =>
+            Some(id :: acc)
+          case _ => None
+        }
+    }
+
```

Comment:
It's not obvious what is the supposed semantics of the extractor. Maybe add a document.